bot/utils/persistent_scheduling.py

```diff
@@ -0,0 +1,155 @@
+        return {task_id: Arrow.utcfromtimestamp(timestamp) for task_id, timestamp in dict_.items()}
+
+    async def reschedule_task(self, task_id: CacheKey) -> None:
+        """Reschedule the task according to `task_factory` and `task_id`."""
+        time = await self.get(task_id)
+        if not time:
+            raise ValueError(f"Task with ID {task_id} was not found for rescheduling.")
+        self._scheduler.schedule_at(time.datetime, task_id, coroutine=self._to_schedule(task_id))
+
+    async def reschedule_all_tasks(self) -> None:
+        """Reschedule all tasks according to `task_factory` and the ID's in the cache."""
+        tasks = await self.cache.items()
+        for task_id, timestamp in tasks:
+            time = Arrow.utcfromtimestamp(timestamp)
+            self._scheduler.schedule_at(time.datetime, task_id, coroutine=self._to_schedule(task_id))
+
+    async def delete(self, task_id: CacheKey) -> None:
+        """
+        Unschedule the task identified by `task_id` and delete it from the cache.
+
+        `cancel` is left as-is in order to be able to pause execution until the next reschedule.
+        Log a warning if the task doesn't exist.
+        """
+        await self.cache.delete(task_id)
+
+        await self._reschedule_task  # Wait for rescheduling to finish to avoid race conditions.
+        self._scheduler.cancel(task_id)
+
+    async def delete_all(self) -> None:
+        """
+        Unschedule all known tasks and wipe the cache.
+
+        `cancel_all` remains untouched so that it can be used in instances such as cog unloads while preserving cache.
+        """
+        self._reschedule_task.cancel()
+        await self.cache.clear()
+
+        self._scheduler.cancel_all()
+
+    async def schedule_at(self, time: datetime, task_id: CacheKey) -> None:
+        """
+        Schedule `coroutine` to be executed at the given `time` and add it to the cache.
+
+        If `time` is timezone aware, then use that timezone to calculate now() when subtracting.
+        If `time` is naïve, then use UTC.
+
+        If `time` is in the past, schedule `coroutine` immediately.
+
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
+        """
+        self._scheduler.schedule_at(time, task_id, self._to_schedule(task_id))
+
+        if time.tzinfo is None:
+            time.replace(tzinfo=timezone.utc)
+        await self.cache.set(task_id, time.timestamp())
+
+    async def schedule_later(self, delay: t.Union[int, float], task_id: CacheKey) -> None:
+        """
+        Schedule `coroutine` to be executed after the given `delay` number of seconds and add it to the cache.
+
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
```

Comment:
I think the closing coroutine part is not relevant for this class since it relies on a factory rather than directly being given a coroutine.